File: src/ui/select.ts
/**
 * 自绘 ↑↓ 选择菜单（对齐现版 Select-Menu）。
 *
 * - ↑↓ 导航（跳过 '' 分隔空行）、数字键直选、Enter 确认、q/Esc 取消、Ctrl+C 退出。
 * - 原地重绘（光标上移 + 清屏到底 + 隐藏光标）→ 不闪烁。
 * - 可选就地排序：传 onMove + movableCount，Shift+↑↓ 或 PgUp/PgDn 在顶部前 N 项内移动选中项。
 * - 非交互/无 TTY 时回退到「打印列表 + 读一行序号」。
 *
 * 注：文本输入（含中文）不在这里——那走 ui/text.ts（raw readValue / cooked readText，后者兼容输入法，评审④）。
 */
import { createInterface, emitKeypressEvents, type Key } from 'node:readline';

import { T } from '../i18n/index.js';
import { CLEAR_DOWN, CLEAR_SCREEN, CR, cursorUp, HIDE_CURSOR, paint, SHOW_CURSOR, type Color } from '../utils/ansi.js';
import { truncateDisplay } from '../utils/display.js';

export interface SelectOptions {
  title?: string;
  /** 菜单项；'' = 不可选的分隔空行（导航跳过）。 */
  items: string[];
  hint?: string;
  /** 顶部绿色 toast（如「已设为默认」），显示一轮。 */
  status?: string;
  /** 初始选中项（记忆选中）。 */
  start?: number;
  /** 按索引上色（如「＋ 新增配置」用黄色）。 */
  colors?: Record<number, Color>;
  /** 顶部可排序区的项数（仅前 N 项可被 Shift+↑↓/PgUp/PgDn 移动）。 */
  movableCount?: number;
  /** 排序回调：交换数据并返回重建后的菜单项标签数组。 */
  onMove?: (from: number, to: number) => string[];
}

/** 返回选中索引；取消（q/Esc/非法）返回 -1。 */
export async function selectMenu(opts: SelectOptions): Promise<number> {
  const stdin = process.stdin;
  const stdout = process.stdout;
  let items = opts.items.slice();

  if (!stdin.isTTY || !stdout.isTTY) return fallbackSelect(opts, items);

  const nextSel = (i: number, d: number): number => {
    const n = items.length;
    do {
      i = (i + d + n) % n;
    } while (items[i] === '');
    return i;
  };
  let idx = opts.start ?? 0;
  if (idx < 0 || idx >= items.length || items[idx] === '') idx = nextSel(Math.max(0, Math.min(idx, items.length - 1)), 1);

  emitKeypressEvents(stdin);
  const wasRaw = stdin.isRaw ?? false;
  stdin.setRawMode(true);
  stdin.resume();
  stdout.write(CLEAR_SCREEN + HIDE_CURSOR); // 清屏归位，制造「整页」感（每进一级菜单都是新页）

  let prevLines = 0;
  const render = (): void => {
    const cols = stdout.columns ?? 80;
    const lines: string[] = [''];
    if (opts.title) {
      lines.push(`  ${paint(opts.title, 'cyan')}`, '');
    }
    if (opts.status) {
      lines.push(`  ${paint(opts.status, 'green')}`, '');
    }
    for (let i = 0; i < items.length; i++) {
      const it = items[i] ?? '';
      if (it === '') {
        lines.push('');
        continue;
      }
      if (i === idx) lines.push(paint(`   ▶ ${it}`, 'green'));
      else lines.push(paint(`     ${it}`, opts.colors?.[i] ?? 'none'));
    }
    lines.push('');
    if (opts.hint) lines.push(`  ${paint(opts.hint, 'dim')}`);

    const block = lines.map((l) => truncateDisplay(l, cols - 1)).join('\n');
    const prefix = prevLines > 0 ? cursorUp(prevLines - 1) + CR + CLEAR_DOWN : '';
    stdout.write(prefix + block);
    prevLines = lines.length;
  };

  return new Promise<number>((resolve) => {
    const cleanup = (result: number): void => {
      stdin.off('keypress', onKey);
      stdout.write(`${SHOW_CURSOR}\n`);
      if (!wasRaw) stdin.setRawMode(false);
      stdin.pause();
      resolve(result);
    };
    const onKey = (str: string | undefined, key: Key): void => {
      if (key?.ctrl && key.name === 'c') {
        cleanup(-1);
        process.exit(130);
      }
      // 就地排序（仅可排序区内）
      if (opts.onMove && opts.movableCount) {
        const up = key?.name === 'pageup' || (Boolean(key?.shift) && key?.name === 'up');
        const down = key?.name === 'pagedown' || (Boolean(key?.shift) && key?.name === 'down');
        if (up && idx > 0 && idx < opts.movableCount) {
          items = opts.onMove(idx, idx - 1);
          idx -= 1;
          render();
          return;
        }
        if (down && idx < opts.movableCount - 1) {
          items = opts.onMove(idx, idx + 1);
          idx += 1;
          render();
          return;
        }
      }
      switch (key?.name) {
        case 'up':
          idx = nextSel(idx, -1);
          render();
          return;
        case 'down':
          idx = nextSel(idx, 1);
          render();
          return;
        case 'return':
        case 'enter':
          cleanup(idx);
          return;
        case 'escape':
          cleanup(-1);
          return;
        default:
          break;
      }
      const ch = str ?? '';
      if (/^[0-9]$/.test(ch)) {
        const n = Number.parseInt(ch, 10);
        if (n >= 1 && n <= items.length && items[n - 1] !== '') cleanup(n - 1);
        return;
      }
      if (ch === 'q') cleanup(-1);
    };
    stdin.on('keypress', onKey);
    render();
  });
}

/** 非交互回退：打印列表 + 读一行序号。 */
async function fallbackSelect(opts: SelectOptions, items: string[]): Promise<number> {
  const stdout = process.stdout;
  stdout.write('\n');
  if (opts.title) stdout.write(`  ${opts.title}\n\n`);
  items.forEach((it, i) => {
    if (it !== '') stdout.write(`   ${i + 1}. ${it}\n`);
  });
  const rl = createInterface({ input: process.stdin, output: process.stdout });
  const ans = await new Promise<string>((res) => rl.question(`  ${T('menu.prompt')}`, res));
  rl.close();
  const t = ans.trim();
  if (t === 'q') return -1;
  if (/^\d+$/.test(t)) {
    const n = Number.parseInt(t, 10);
    if (n >= 1 && n <= items.length && items[n - 1] !== '') return n - 1;
  }
  return -1;
}

File: internal/tui/select.go
package tui

import (
	"os"
	"strconv"
	"strings"

	"golang.org/x/term"

	"github.com/becomeless/cc-x/internal/display"
	"github.com/becomeless/cc-x/internal/i18n"
)

// SelectOptions 配置一次 ↑↓ 选择菜单。” 项为不可选分隔空行（导航跳过）。
type SelectOptions struct {
	Title        string
	Notice       string // 标题下方黄字横幅（如「有新版本」），常驻显示
	Items        []string
	Hint         string
	Status       string        // 顶部绿色 toast
	Start        int           // 初始选中（记忆选中）
	Colors       map[int]Color // 按索引上色
	MovableCount int           // 顶部可排序区项数
	OnMove       func(from, to int) []string
	NoNumber     bool // 关闭行首序号（默认显示，与数字直选一致；编辑表单项多于 9 个时关闭）
}

// SelectMenu 自绘 ↑↓ 选择菜单，返回选中索引；取消（q/Esc/非法）返回 -1；Ctrl+C 恢复终端后以 130 退出。
// 自管 raw 进出（与 ReadText 的 cooked 互斥，同一时刻只跑一套）。非 TTY 回退到打印列表 + 读序号。
func SelectMenu(t *Terminal, opts SelectOptions) int {
	if !t.IsTTY() {
		return fallbackSelect(t, opts)
	}
	items := append([]string(nil), opts.Items...)
	nextSel := func(i, d int) int {
		n := len(items)
		for {
			i = ((i+d)%n + n) % n
			if items[i] != "" {
				return i
			}
		}
	}
	idx := opts.Start
	if idx < 0 || idx >= len(items) || items[idx] == "" {
		c := idx
		if c < 0 {
			c = 0
		}
		if c > len(items)-1 {
			c = len(items) - 1
		}
		idx = nextSel(c, 1)
	}

	if err := t.MakeRaw(); err != nil {
		return fallbackSelect(t, opts)
	}
	t.Write(ClearScreen + HideCursor)

	prevLines := 0
	render := func() {
		lines := []string{""}
		if opts.Title != "" {
			lines = append(lines, "  "+Paint(opts.Title, ColorCyan), "")
		}
		if opts.Notice != "" {
			lines = append(lines, "  "+Paint(opts.Notice, ColorYellow), "")
		}
		if opts.Status != "" {
			lines = append(lines, "  "+Paint(opts.Status, ColorGreen), "")
		}
		for i, it := range items {
			if it == "" {
				lines = append(lines, "")
				continue
			}
			num := ""
			if !opts.NoNumber {
				num = strconv.Itoa(i+1) + ". " // 行首序号，与数字键直选对应
			}
			if i == idx {
				lines = append(lines, Paint("   ▶ "+num+it, ColorGreen))
			} else {
				col := ColorNone
				if c, ok := opts.Colors[i]; ok {
					col = c
				}
				lines = append(lines, Paint("     "+num+it, col))
			}
		}
		lines = append(lines, "")
		if opts.Hint != "" {
			lines = append(lines, "  "+Paint(opts.Hint, ColorDim))
		}

		cols := termWidth(t)
		var b strings.Builder
		if prevLines > 0 {
			b.WriteString(CursorUp(prevLines-1) + CR + ClearDown)
		}
		for i, l := range lines {
			if i > 0 {
				b.WriteByte('\n')
			}
			b.WriteString(display.Truncate(l, cols-1))
		}
		t.Write(b.String())
		prevLines = len(lines)
	}
	cleanup := func() {
		t.Write(ShowCursor + "\n")
		t.Restore()
	}

	render()
	for {
		k := t.ReadKey()
		switch k.Type {
		case KeyCtrlC:
			cleanup()
			os.Exit(130)
		case KeyUp:
			idx = nextSel(idx, -1)
			render()
		case KeyDown:
			idx = nextSel(idx, 1)
			render()
		case KeyShiftUp, KeyPgUp:
			if opts.OnMove != nil && idx > 0 && idx < opts.MovableCount {
				items = opts.OnMove(idx, idx-1)
				idx--
				render()
			}
		case KeyShiftDown, KeyPgDn:
			if opts.OnMove != nil && idx < opts.MovableCount-1 {
				items = opts.OnMove(idx, idx+1)
				idx++
				render()
			}
		case KeyEnter:
			cleanup()
			return idx
		case KeyEsc:
			cleanup()
			return -1
		case KeyDigit:
			if !opts.NoNumber {
				n := int(k.Rune - '0')
				if n >= 1 && n <= len(items) && items[n-1] != "" {
					cleanup()
					return n - 1
				}
			}
		case KeyChar:
			if k.Rune == 'q' {
				cleanup()
				return -1
			}
		}
	}
}

func termWidth(t *Terminal) int {
	w, _, err := term.GetSize(int(t.Out.Fd()))
	if err != nil || w <= 0 {
		return 80
	}
	return w
}

// fallbackSelect 非交互回退：打印列表 + 读一行序号。
func fallbackSelect(t *Terminal, opts SelectOptions) int {
	t.Write("\n")
	if opts.Title != "" {
		t.Write("  " + opts.Title + "\n\n")
	}
	for i, it := range opts.Items {
		if it != "" {
			t.Write("   " + strconv.Itoa(i+1) + ". " + it + "\n")
		}
	}
	ans, ok := t.ReadLine("  " + i18n.T("menu.prompt"))
	if !ok {
		return -1
	}
	s := strings.TrimSpace(ans)
	if s == "q" {
		return -1
	}
	if n, err := strconv.Atoi(s); err == nil && n >= 1 && n <= len(opts.Items) && opts.Items[n-1] != "" {
		return n - 1
	}
	return -1
}
